notebook/services/contents/filemanager.py

```diff
@@ -489,19 +489,19 @@ def delete_file(self, path):
         path = path.strip('/')
         os_path = self._get_os_path(path)
         rm = os.unlink
-        if os.path.isdir(os_path):
-            listing = os.listdir(os_path)
-            # Don't delete non-empty directories.
-            # A directory containing only leftover checkpoints is
-            # considered empty.
-            cp_dir = getattr(self.checkpoints, 'checkpoint_dir', None)
-            for entry in listing:
-                if entry != cp_dir:
-                    raise web.HTTPError(400, u'Directory %s not empty' % os_path)
-        elif not os.path.isfile(os_path):
-            raise web.HTTPError(404, u'File does not exist: %s' % os_path)
+        if not os.path.exists(os_path):
+            raise web.HTTPError(404, u'File or directory does not exist: %s' % os_path)
 
         if self.delete_to_trash:
+            if os.path.isdir(os_path):
+                listing = os.listdir(os_path)
+                # Don't send non-empty directories to trash.
```

Comment:
I think you've got this backwards. The idea is that if we're using trash (if `self.delete_to_trash` is true), then we *can* delete non-empty directories, because the user can get them back from trash. If we're *not* using trash, then deleting is permanent, so we should still refuse to delete non-empty directories.